include/envoy/server/overload/proactive_resource_monitor.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/common/pure.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Server {
+
+class ProactiveResourceMonitor {
+public:
+  ProactiveResourceMonitor() = default;
+  virtual ~ProactiveResourceMonitor() = default;
+  /**
+   * Tries to allocate resource for given resource monitor in thread safe manner.
+   * Returns true if there is enough resource quota available and allocation has succeeded, false
+   * otherwise.
+   * @param increment to add to current resource usage value and compare against configured max
+   * threshold.
+   */
+  virtual bool tryAllocateResource(uint64_t increment) PURE;
+  /**
+   * Tries to deallocate resource for given resource monitor in thread safe manner.
+   * Returns true if there is enough resource quota available and deallocation has succeeded, false
+   * otherwise.
+   * @param decrement to subtract from current resource usage value.
+   */
+  virtual bool tryDeallocateResource(uint64_t decrement) PURE;
+  /**
+   * Returns current resource usage tracked by monitor.
+   */
+  virtual uint64_t currentResourceUsage() const PURE;
+  /**
+   * Returns max resource usage configured in monitor.
+   */
+  virtual uint64_t maxResourceUsage() const PURE;
+};
+
+using ProactiveResourceMonitorPtr = std::unique_ptr<ProactiveResourceMonitor>;
+
+// Example of proactive resource monitor. To be removed.
+class ActiveConnectionsResourceMonitor : public ProactiveResourceMonitor {
+public:
+  ActiveConnectionsResourceMonitor(uint64_t max_active_conns)
+      : max_(max_active_conns), current_(0){};
+
+  bool tryAllocateResource(uint64_t increment) {
+    int64_t new_val = (current_ += increment);
```

Comment:
@antoniovicente i still feel like there is race condition here. Let's say we compute value on line 52, then another thread deallocates the resource (so that current usage now can accommodate for resource allocation in current thread) but current call would fail as if there is not enough quota to allocate resource. I think this code can be improved like:
```
  bool tryAllocateResource(uint64_t increment) {
    current_ += increment;
    if (current_ > max_ || current_ < 0) {
     //Race condition can occur between the lines here.
      current_ -= increment;
      return false;
    }
    return true;
  }
```
But still does not fully eliminate race probability.